### ISSUE TITLE
blkdiscard all unpartitioned space

### ISSUE THREAD

**Issue by rdiez** [util-linux/util-linux#2804]
Could you add an option to TRIM all unpartitioned space on a disk?

Some background information is here:

https://unix.stackexchange.com/questions/97143/utility-to-trim-unallocated-space-on-drive

**Comment by karelzak** [comment]
blkdiscard provides `--offset` and `--length` to address the area to discard and `sfdisk --list-free` prints unpartitioned areas. All you need is to use it together :-) Yes, it's pretty dangerous.

I'm not sure if we want to link blkdiscard for this purpose. It sounds like an unusual use case.

**Comment by rdiez** [comment]
Linux only automatically trims mounted partitions. If you clone an SSD with Clonezilla, or write an image with dd to an SD card for a Raspberry Pi, any old data not overwritten will never be trimmed. If you delete a partition with gnome-disks or the like, I doubt it would get trimmed. For example, sfdisk can delete partitions, but it does not mention trimming at all. It feels a little silly leaving that opportunity unused. Even creating a partition to cover all empty space will probably not trim it until the global systemd fstrim.timer kicks in, perhaps next week.

If blkdiscard could automatically trim unused space and on demand, I am guessing that many people would use it after running dd or the like. Otherwise, most people won't bother, which would be a shame in terms of performance and/or wear leveling.

You could argue that you should trim the whole disk beforehand, but trim is often only an afterthought when performance suffers, or you may get the disk from somebody else who didn't. Besides, you may only be writing or shrinking one partition.

Do you know of any tool or script to automatically trim unpartitioned space? Even though sfdisk is touted as "script-oriented", it doesn't seem like a 5-minute job.

**Comment by karelzak** [comment]
What about adding `---trim` to fdisks? It has all the information (offsets & sizes of the areas). We can use it, for example, when you delete a partition or on demand when you create a new partition. Would it make more sense in fdisks?

**Comment by rdiez** [comment]
I don't really mind which tool. I don't have a good overview anyway.

For me, the main thing would be the ability to trim the unpartitioned space. Otherwise, you would have to create and delete partitions just for that purpose, which is a pain, especially if there are holes (several of them).

But trimming deleted or newly-created partitions is a good idea too.

I just realised that this is not just a thing for real disks or even some SD cards: apparently QEMU also honours TRIM commands and releases space in .qcow2 virtual disk files.

**Comment by karelzak** [comment]
OK, I'll think about it for the fdisks, including unpartitioned space.

**Comment by rdiez** [comment]
First of all, many thanks for implementing this!

The 'U' command is what I would use. The trouble is, I would then have to manually run this command for all unallocated areas, one by one.

A command-line option like "--trim-unpartitioned" would be much more comfortable and safer for users like me. Or is such an option already there but I have missed it?

By the way, there is a typo in message "has unspeficied range"

**Comment by karelzak** [comment]
I'll think about --trim-unpartitioned for sfdisk :-)

**Comment by rdiez** [comment]
Among other things, I am trying to automate writing an SD card image for a Raspberry Pi. The SD cards often get reused, so the idea is that you write the sectors or partitions you need and then trim the rest. Another usage scenario is to trim existing SD cards, so that the system will run faster because most of the space is already trimmed. Yet another scenario is to automatically trim QEMU images from inside the guest which have been created and used without taking Trim into consideration. That is why a script-friendly solution would be nice indeed.

Besides, if fdisk can trim unused space, and sfdisk cannot, then it looks a little silly, doesn't it? };-)

**Comment by feignwe** [comment]
> blkdiscard provides --offset and --length to address the area to discard and sfdisk --list-free prints unpartitioned areas. All you need is to use it together :-) Yes, it's pretty dangerous.

@karelzak 
> [sfdisk: add --discard-free](https://github.com/karelzak/util-linux-work/commit/e13e47207670cd20ff6c2241544ac6d9660d4c31)

I tried `sfdisk --discard-free` and it deleted all logical partitions inside one extended partition from the table, except the first one. 
The actual data inside the partitions remained intact, so I simply recreated the partitions with the same initial/final sector numbers and it worked. 
I think it is discarding the Extended Boot Records of extended partitions, which actually resides in "unallocated" space (space that can potentially be used to create new partitions).
For example, `sda1` is primary, `sda2` is extended, `sda5` is the first logical inside `sda2`, `sda6` is the second logical inside `sda2`, and so on. It deleted `sda6`, `sda7`, `sda8` and `sda9`, but not `sda5`.
Master boot record, sectors from 0 to 2047 seemed to have remained intact.

My version is: `sfdisk from util-linux 2.41.1`

I forgot to mention, the partition table type is `MBR`.
